config/runtime.exs

```diff
@@ -1,36 +1,24 @@
 import Config
+import Trento.Config
 
 # config/runtime.exs is executed for all environments, including
 # during releases. It is executed after compilation and before the
 # system starts, so it is typically used to load production configuration
 # and secrets from environment variables or elsewhere. Do not define
 # any compile-time configuration in here, as it won't be applied.
-# The block below contains prod specific runtime configuration.
-if config_env() == :prod do
```

Comment:
I believe the intent here is to raise errors when some critical config is missing when running the application in `:prod`.
And I personally would keep that semantic instead of letting the app use arbitrarily default set by us.
When you run the app in prod mode, that is your complete responsibility to provide the correct config/env imho.

In addition this is the community standard afaiu :smile:

Reply:
The problem `MIX_ENV` is giant switch with a cascading side-effect on many other parameters at once, both at compile-time and at run-time, and as I've always vocally argued that this is a huge operational anti-pattern, very common in many major frameworks (Rails, Django, Symfony, they all share this outdated macro-config-switch-of-doom approach, and I was underwhelmed to see this in Elixir as well, but I guess they were inspired by Rails 😬).

The "environment" as a concept should never be distilled into a single variable; it's like a value object: it's identified by the all the values it's composed by, and it has no other identity than the sum of all those.
IOW, having a "development environment" means having a bunch of runtime configuration values that are used during development, and that's it. There shouldn't be a `dev` environment; in fact having a "development mode single switch" that flips many settings at once is a huge PITA, because it couples configuration values with a compounded semantic that might should instead be independently controllable for each value.

Let me make a practical example: there is nothing about using `localhost` that tells you it's production, development or testing. If people decided to run multiple containers in the same network namespace, because reasons, then they would use `localhost`, and whether that's production or not is not something we can predict.
Assuming the environment from the configuration values is a backwards model that has origin in the age of fully local development, when we could afford assumptions about how people ran things. This is no longer a safe assumption to make: all kinds of environments could be distributed, nowadays. In fact you could even do distributed development and then monolithic deployments!

The real reason of `MIX_ENV` to exist is to determine how the sources are built, rather than how they are ran. This is why `runtime.exs` was only relatively recently [introduced in Elixir](https://elixir-lang.org/blog/2020/10/06/elixir-v1-11-0-released/#configruntimeexs-and-mix-appconfig).


> I believe the intent here is to raise errors when some critical config is missing when running the application in :prod.

That is still the case, but the only critical config now raising due to missing value is `"SECRET_KEY_BASE"`, because it's not secure to provide a default there.
All the other ones are fine: using wrong values will result in an error, so if the defaults are wrong the get an error, as they should be.


> When you run the app in prod mode, that is your complete responsibility to provide the correct config/env imho.

That is still the case, but see the point above about not making assumptions about what is correct or not in any given environment. The validation logic should be the same everywhere.


> In addition this is the community standard afaiu smile

FWIW I did some research but I couldn't find an established convention about how to write configs; the way it's done is entirely dependant on each single project.

A lot of projects only limit themselves to a single `config.exs` file for compile-time settings.
Some people scatter things across files without logic and even repeat the same `System.get_env()` call with the same fallback value in every single file. 🙄 
I took inspiration from https://github.com/plausible/analytics. @dottorblaster agreed it looked like a good way forward.

Maybe @fabriziosestito will shed some light on this.

Reply:
Fair enough, interesting perspective, thanks for sharing :rocket:
With this addition and Fabrizio's comment I believe we can find a middle ground to go.